File: assignment9/word.cpp
#include<iostream>
#include <unordered_map>
#include <map>
#include<vector>
#include<string>
//#include<thread>
//#include<mutex>
using std::unordered_map;
using std::map;
using std::vector;
using std::string;
//using std::thread;
//using std::mutex;
using std::cout;
using std::endl;
class Node
{
//0 top, 1 tr, 2 right, 3 rb, 4 bottom, 5 bl, 6 left,7  lt
public:
    vector<int> neighbors;
    char index;
    bool isValid=true;
    Node(char c){
        this->neighbors.resize(8);
        this->index=c;
        for(auto & e : neighbors){
            e=-1;
        }
    }
};
class Grid{
private:
public:
    int colLength = 0;
    int rowHeight=0;
    vector<Node> nodes;
    Grid(string gridString){
        int row=0,col=0;
        for(auto c : gridString){
            if(c!='\n')nodes.push_back(Node(c));
        }
        for(int i=0;i<gridString.size();i++){
            if (gridString.at(i) != '\n') {
                //create links between nodes
                if(col!=0){
                    nodes.at(row*colLength + col-1).neighbors.at(2)=i - row;//set previous link to right
                    nodes.at(i-row).neighbors.at(6) = row*colLength + col-1;//set this node link to left
                }
                if(row!=0){
                    nodes.at((row-1)*colLength + col).neighbors.at(4)=i - row;
                    nodes.at(i - row).neighbors.at(0)=(row-1)*colLength + col;//top link
                    if(col<colLength - 1){
                        nodes.at((row-1)*colLength+col + 1).neighbors.at(5)=i-row;
                        nodes.at(i-row).neighbors.at(1)=(row-1)*colLength  + col+1;//top right link
                    }
                }
                if(col!=0&&row!=0){
                    nodes.at((row-1)*colLength + col-1).neighbors.at(3)=i-row;
                    nodes.at(i-row).neighbors.at(7)=(row-1)*colLength + col-1;//top left link
                }
                col++;
            }
            else{
                this->colLength = col;
                col=0;
                row++;
            }
        }
        this->rowHeight = row;
    }
    void UnlinkInValid(){
        /*
        for(int i=nodes.size()-1;i>=0; i --){
            if(nodes.at(i).isValid==false){
                auto temp_node = nodes.at(i);
                auto temp_node_top = nodes.at(i).neighbors.at(0);
                while(!nodes[temp_node_top].isValid)
            }
        }*/
        for(int c=0;c<colLength;c++){
            auto temp_one_col = vector<char>();
            for(int r=0; r<rowHeight; r++){
                auto node=nodes.at(r*colLength + c);
                if(node.isValid)
                temp_one_col.push_back(node.index);
            }
            for(int r=rowHeight - 1; r>=0; r--){
                if(temp_one_col.size()>0){
                    nodes.at(r*colLength + c).index = temp_one_col.back();
                    nodes.at(r*colLength + c).isValid = true;
                    temp_one_col.pop_back();
                }
                else{
                    nodes.at(r*colLength + c).isValid = false;
                }

            }
        }
    }
    Grid* fork() const{
         return new Grid(*this);
    }
};
class TireNode{
private:
public:
    unordered_map<char,TireNode*> value;
    bool isWordEnd=false;
    void Append(string word){
        auto c = word.front();
        if(word.size()==0){
            this->isWordEnd=true;
        }
        else{
            word.erase(word.begin());
            if(value.find(c)!=value.end()){//find it
                value.at(c)->Append(word);
            }
            else{
                value[c] = new TireNode();
                value[c]->Append(word);
            }
        }
    }
};
class AnswerTree{
public:
    string value;
    vector<AnswerTree> children;
    AnswerTree(string v){
        value = v;
    }
};
//tried to use multi-threading but it seems i have to add -lpthread while compile
/*
mutex locker;
void grid_search_atom(const Grid & grid, int startIndex, vector<Grid> * result_grid, vector<string> * result_string, int* threadCount){
    string resultString="";
    



    locker.lock();
    *threadCount+=1;
    locker.unlock();

}*/
/*
void grid_search(Grid & grid, const string & word, const TireNode& tire){
    int threadCount =0;
    vector<Grid> result_grid;
    vector<string> result_string;
    /*
    for(int i=0;i<grid.nodes.size();i++){
        //thread t(grid_search_atom, grid, i, &result_grid, &result_string, & threadCount);
        //t.detach();
        
    }
    while(threadCount!=grid.nodes.size()-1);
    *
}*/
void testFunction_PrintGrid(Grid & g){
    for(int i=0;i<g.nodes.size();i++){
        auto node = g.nodes.at(i);
        cout<<node.index<<((node.isValid)?" ":"*");
        if(i%g.rowHeight==g.rowHeight - 1)
        cout<<'\n';
    }
}
void getAnswer(vector<AnswerTree> & result, vector<string>* answer){
    
}
void grid_search_atom(Grid grid, int startIndex, vector<Grid> * result_grid, vector<string> * result_string,const TireNode * tire, string currentString){
    auto current_Node = grid.nodes.at(startIndex);
    grid.nodes.at(startIndex).isValid=false;//set search flag
    if(tire->value.find(current_Node.index)==tire->value.end()){
        //find failed
        return;
    }else{
        tire = tire->value.at(current_Node.index);
    }
    currentString  += current_Node.index;
    if(tire->isWordEnd){
        result_string->push_back(currentString);
        result_grid->push_back(grid);
        return;
    }
    for(auto & neighbourIndex : current_Node.neighbors){
        if(neighbourIndex!=-1){//check if neighbor is exist
            auto neigbour = grid.nodes.at(neighbourIndex);
            if(neigbour.isValid==true){//un searched;
                //then go searching
                grid_search_atom(grid,neighbourIndex,result_grid,result_string, tire,currentString);
            }
        }
    }
}
void search_grid(Grid & g, vector<TireNode*> & tire,vector<AnswerTree>* result,int currentDepth){
    vector<Grid>  result_grid;
     vector<string>  result_string;
    for(int i=0; i<g.nodes.size();i++){
        if(g.nodes.at(i).isValid){
            grid_search_atom(g,i,&result_grid,&result_string,tire.at(currentDepth),"");
        }
    }
    for(int i=0; i < result_grid.size();i++){
        result_grid.at(i).UnlinkInValid();
        result->push_back(AnswerTree(result_string.at(i)));
        if(currentDepth != tire.size() - 1){
            search_grid(result_grid.at(i),tire,&(result->back().children),currentDepth + 1);
        }
    }
}
int main(){
    auto a = TireNode();
    a.Append("post");
    a.Append("pots");
    a.Append("stop");
    a.Append("opts");
    auto b = TireNode();
    b.Append("bedroom");
    b.Append("vampire");
    b.Append("vitamin");
    auto g = Grid("vanmo\nipveo\ntoarr\ntsmed\nmiipb\n");
    vector<TireNode*> tireList;
    tireList.push_back(&a);
    tireList.push_back(&b);
    tireList.push_back(&b);
    tireList.push_back(&b);
    vector<AnswerTree> result;
    search_grid(g,tireList,&result,0);
    vector<string> answer;
    getAnswer(result,&answer);
    //testFunction_PrintGrid(g);
    /*
    vector<Grid> result_grid;
    vector<string> result_string;
    for(int i=0;i<g.nodes.size();i++){
        if(g.nodes.at(i).isValid){
            grid_search_atom(g,i,&result_grid,&result_string,&a,"");
        }
    }

    testFunction_PrintGrid(result_grid.at(0));
    cout<<endl;
    result_grid.at(0).UnlinkInValid();
    auto temp_grid = result_grid.at(0);
    result_grid.clear();
    result_string.clear();
    for(int i=0;i<g.nodes.size();i++){
        if(g.nodes.at(i).isValid){
            grid_search_atom(temp_grid,i,&result_grid,&result_string,&b,"");
        }
    }
    testFunction_PrintGrid(result_grid.at(0));
    /*
    for(int i=0; i < result_string.size();i++){
        cout<<result_string.at(i)<<endl;
        testFunction_PrintGrid(result_grid.at(i));
    }*/
}

File: assignment9/word.py
import copy
from sys  import argv
import sys
class Node:
#0 top, 1 tr, 2 right, 3 rb, 4 bottom, 5 bl, 6 left,7  lt
    def __init__(self,c):
        self.neighbors = []
        self.index=''
        self.isValid=True
        self.index=c
        for i  in range(8):
            self.neighbors.append(-1)
    pass
class Grid:
    def __init__(self,gridString):
        self.colLength = 0
        self.rowHeight=0
        self.nodes=[]

        row=0
        col=0
        for c in gridString:
            if(c!='\n'):
                self.nodes.append(Node(c))
        for i in range(gridString.__len__()):
            if (gridString[i] != '\n'):
                #create links between self.nodes
                colLength = self.colLength
                if(col!=0):
                    self.nodes[row*colLength + col-1].neighbors[2]=i - row#set previous link to right
                    self.nodes[i-row].neighbors[6] = row*colLength + col-1#set this node link to left
                if(row!=0):
                    self.nodes[(row-1)*colLength + col].neighbors[4]=i - row
                    self.nodes[i - row].neighbors[0]=(row-1)*colLength + col#top link
                    if(col<colLength - 1):
                        self.nodes[(row-1)*colLength+col + 1].neighbors[5]=i-row
                        self.nodes[i-row].neighbors[1]=(row-1)*colLength  + col+1#top right link
                if(col!=0 and row!=0):
                    self.nodes[(row-1)*colLength + col-1].neighbors[3]=i-row
                    self.nodes[i-row].neighbors[7]=(row-1)*colLength + col-1#top left link
                col+=1
            else:
                self.colLength = col
                col=0
                row+=1
        self.rowHeight = row
    def UnlinkInValid(self):
        colLength = self.colLength
        rowHeight = self.rowHeight
        for c in range(colLength):
            temp_one_col = []
            for  r in range(rowHeight):
                node=self.nodes[r*colLength + c]
                if(node.isValid):
                    temp_one_col.append(node.index)
            for r in range (rowHeight -1, -1,-1):
                if(temp_one_col.__len__()>0):
                    self.nodes[r*colLength + c].index = temp_one_col[temp_one_col.__len__() - 1]
                    self.nodes[r*colLength + c].isValid = True
                    temp_one_col.pop()
                else:
                    self.nodes[r*colLength + c].isValid = False
class TireNode(object):
    #unordered_map<char,TireNode*> value
    def __init__(self):
        self.value = {}
        self.isWordEnd=False
        pass
    def Append(self, word):
        if(word.__len__()==0):
            self.isWordEnd=True
        else:
            c = word[0]
            word = word[1:]
            #word.erase(word.begin())
            if(self.value.__contains__(c)):#find it
                self.value[c].Append(word)
            else:
                self.value[c] = TireNode()
                self.value[c].Append(word)
class AnswerTree:
    def __init__(self, v = ""):
        self.value = v
        self.children = []

def testFunction_PrintGrid(g:Grid):
    for i in range(g.nodes.__len__()):
        node = g.nodes[i]
        if(node.isValid):
            validFlag = " "
        else:
            validFlag = "*"
        print(node.index + validFlag,end="")
        if(i%g.rowHeight==g.rowHeight - 1):
            print()
    
def grid_search_atom(g, startIndex:int,result_grid,result_string,tire, currentString=""):
    grid = copy.deepcopy(g)
    startIndex = int(startIndex)
    current_Node = grid.nodes[startIndex]
    grid.nodes[startIndex].isValid=False#set search flag
    if(not (current_Node.index in tire.value)):
        #find failed
        return 
    else:
        newtire = tire.value[current_Node.index]
    currentString  += current_Node.index
    if(newtire.isWordEnd):
        result_string.append(currentString)
        result_grid.append(copy.deepcopy(grid))
        return
    for neighbourIndex in current_Node.neighbors :
        if(neighbourIndex!=-1):#check if neighbor is exist
            neigbour = grid.nodes[int(neighbourIndex)]
            if(neigbour.isValid==True):#un searched
                #then go searching
                grid_search_atom(grid,neighbourIndex,result_grid,result_string, newtire,currentString)
                
def search_grid(g:Grid, tire,result, currentDepth):
    result_grid = []
    result_string = []
    for i  in range(g.nodes.__len__()):
        if(g.nodes[i].isValid):
            grid_search_atom(g,i,result_grid,result_string,tire[currentDepth],"")
    for  i in range(result_grid.__len__()):
        result_grid[i].UnlinkInValid()
        result.append(AnswerTree(result_string[i]))
        if(currentDepth != tire.__len__()- 1):
            search_grid(result_grid[i],tire,result[result.__len__() - 1].children,currentDepth + 1)
def get_answerList(result,words,resultList,currentString, depth):
    pass
def get_tires(file_name,words):
    tires = []
    word_lengths={}
    for word in words:
        word_lengths[word.__len__()] = TireNode()
    #small first
    f = open(file_name,'r')
    line = f.readline()
    while line:
        line_length = line.__len__() - 1
        if(  line_length in word_lengths):
            word_lengths[line_length].Append(line[0:line_length])
        line = f.readline()
    f.close()

    for word in words:
        tires.append(word_lengths[word.__len__()])
    return tires
    pass
def search_and_print(grid_string,words,small_list,large_list):
    grid = Grid(grid_string)
    tires = []
    resultTree = []
    resultList = []
    tires = get_tires(small_list,words)
    search_grid(grid, tires,resultTree,0)

    #get_answerList(resultTree,words,resultList)
    for answer in resultList:
        print(answer)
    
    if(resultList.__len__()==0):#cannot find
        tires = get_tires(large_list,words)
        search_grid(grid, tires,resultTree,0)
        resultList = get_answerList(resultTree,words)
        for answer in resultList:
            print(answer)
    #check & generate
    pass
if __name__ == "__main__":
    # a = TireNode()
    # a.Append("post")
    # a.Append("pots")
    # a.Append("stop")
    # a.Append("opts")
    # b = TireNode()
    # b.Append("bedroom")
    # b.Append("vampire")
    # b.Append("vitamin")
    # g = Grid("vanmo\nipveo\ntoarr\ntsmed\nmiipb\n")
    # tireList = []
    # tireList.append(a)
    # tireList.append(b)
    # tireList.append(b)
    # tireList.append(b)
    # result = []
    # search_grid(g,tireList,result,0)
    small_list = argv[1]
    large_list = argv[2]
    puzzles_lines = sys.stdin.read().split("\n")
    grid_size = 0
    grid_string = ""
    for i in range(puzzles_lines.__len__()):
        words = []
        if(grid_size == 0 ):
            grid_string = ""
            grid_size = puzzles_lines[i].__len__() 
            grid_string += puzzles_lines[i] + "\n"
        else:
            if(grid_size == 1):
                words = puzzles_lines[i].split()
                search_and_print(grid_string,words,small_list,large_list)
            else:
                grid_string += puzzles_lines[i] + "\n"
            grid_size -=1
        #print(words)
        #print(grid_string)
    pass